File: src/debug.zig
const c = @import("./chunk.zig");
const Chunk = c.Chunk(u8);
const OpCode = c.OpCode;
const std = @import("std");
const Value = @import("./value.zig").Value;
const VM = @import("./vm.zig").VM;
const keto = @import("./keto.zig");

pub fn disassembleChunk(self: *Chunk, name: []const u8) void {
    keto.log.info("=== {s} ===\n", .{name});
    var offset: usize = 0;
    while (offset < self.count()) {
        offset = disassembleInstruction(
            self,
            offset,
        );
    }
}

pub fn disassembleInstruction(self: *Chunk, offset: usize) usize {
    keto.log.info("[{x:0>4}] ", .{offset});
    if (offset > 0 and self.lines.items[offset] == self.lines.items[offset - 1]) {
        keto.log.info("    | ", .{});
    } else if (self.lines.items.len > 0) {
        keto.log.info("{d: >5} ", .{self.lines.items[offset]});
    } else {
        keto.log.info("????? ", .{});
    }
    const instruction = self.code.items[offset];
    switch (@intToEnum(OpCode, instruction)) {
        .CONSTANT => return constantInstruction(.CONSTANT, self, offset),
        .NEGATE => return simpleInstruction(.NEGATE, offset),
        .ADD => return simpleInstruction(.ADD, offset),
        .SUBTRACT => return simpleInstruction(.SUBTRACT, offset),
        .MULTIPLY => return simpleInstruction(.MULTIPLY, offset),
        .DIVIDE => return simpleInstruction(.DIVIDE, offset),
        .RETURN => return simpleInstruction(.RETURN, offset),
        _ => {
            keto.log.info("Unknown opcode {x:0>2}\n", .{instruction});
            return offset + 1;
        },
    }
    return offset + 1;
}

pub fn simpleInstruction(opcode: OpCode, offset: usize) usize {
    keto.log.info("{s}\n", .{@tagName(opcode)});
    return offset + 1;
}

pub fn constantInstruction(opcode: OpCode, chunk: *Chunk, offset: usize) usize {
    const constant = chunk.code.items[offset + 1];
    keto.log.info("{s}\t{d: >4}: ", .{ @tagName(opcode), constant });

    printValue(chunk.constants.items[constant]);
    keto.log.info("\n", .{});
    return offset + 2;
}

pub fn printValue(value: Value) void {
    keto.log.info("'{}'", .{value});
}

pub fn printStack(vm: *VM) void {
    keto.log.info("             Stack: ", .{});
    for (vm.stack) |*value| {
        if (@ptrToInt(value) >= @ptrToInt(vm.stackTop.ptr))
            break;
        keto.log.info("[ ", .{});
        printValue(value.*);
        keto.log.info(" ]", .{});
    }
    keto.log.info("\n", .{});
}

File: Keto/Debug.cs
namespace Keto;

public static class Debug
{
    public static void Disassemble(this Chunk chunk, string name)
    {
        Print($"== {name} ==\n");
        for (int offset = 0; offset < chunk.Count;)
        {
            offset = chunk.DisassembleInstruction(offset);
        }
    }

    public static int DisassembleInstruction(this Chunk chunk, int offset)
    {
        Print($"{offset:D4} ");

        if (offset > 0 && chunk.Lines[offset] == chunk.Lines[offset - 1])
        {
            Print("   | ");
        }
        else
        {
            Print($"{chunk.Lines[offset],4} ");
        }

        byte instruction = chunk.Code[offset];

        switch (instruction)
        {
            case (byte)OpCode.Constant:
                return ConstantInstruction(((OpCode)instruction).ToString().ToUpper(), chunk, offset);

            case (byte)OpCode.Nil:
            case (byte)OpCode.True:
            case (byte)OpCode.False:
            case (byte)OpCode.Equal:
            case (byte)OpCode.Greater:
            case (byte)OpCode.Less:
            case (byte)OpCode.Return:
            case (byte)OpCode.Negate:
            case (byte)OpCode.Add:
            case (byte)OpCode.Subtract:
            case (byte)OpCode.Multiply:
            case (byte)OpCode.Divide:
            case (byte)OpCode.Not:
                return SimpleInstruction(((OpCode)instruction).ToString().ToUpper(), offset);
            default:
                Print($"Unknown opcode {instruction:D}\n");
                return offset + 1;
        }
    }

    private static int ConstantInstruction(string name, Chunk chunk, int offset)
    {
        var constant = chunk.Code[offset + 1];
        Print($"{name,-16} {offset:D4} '");
        PrintValue(chunk.Constants[constant]);
        Print("'\n");
        return offset + 2;
    }

    private static int SimpleInstruction(string name, int offset)
    {
        Print($"{name}\n");
        return offset + 1;
    }

    public static void PrintValue(Value value)
    {
        switch (value.Type)
        {
            case ValueType.Bool:
                Print(value.ToBool() ? "true" : "false");
                break;
            case ValueType.Nil:
                Print("nil");
                break;
            case ValueType.Number:
                Print($"{value.ToNumber():G}");
                break;
        }
    }

    public static void Print(string value)
    {
        Console.Write(value);
    }

    public static void PrintError(string value)
    {
        Console.Error.Write(value);
    }
}